posts/2021/01/packaging_painpoints_in_2021.md

```diff
@@ -0,0 +1,219 @@
+
+- Isolated builds with `pyproject.toml` are quite complicated to spec, and still unclear how that will hold up in the real world
+
+_RG: Isolated builds are a good idea, but with ABI compatibility constraints one needs to deal with like when building against the NumPy C API, getting `pyproject.toml` right is complicated. See for example [the `oldest-supported-numpy` meta-package](https://github.com/scipy/oldest-supported-numpy/blob/master/setup.cfg) to help with this._
+
+- Filling out configuration is tedious
+- Which freaking tool should I use!
+- Reproducible builds! (practically impossible) – IPython is reproducible for both `.tar.gz` anf `.zip`, but needs custom post-process.
+
+_RG: See [this blog post from Matthias Bussonnier](https://labs.quansight.org/blog/2020/08/ipython-reproducible-builds/) for more on this topic._
+
+- `setup.py` is dynamic making downstream packaging/automation much harder (including for conda) 
+    - Use [flit](http://flit.readthedocs.org/) – completely static! 
+
+_RG: For pure Python packages Flit is great indeed. For packages with compiled code, we'll be stuck with `setup.py` for a while longer I'm afraid. It is what it is - this has gotten better already with `pyproject.toml` and it's not one of the biggest issues anymore imho._
+
+- Lack of relocation tools in Windows, i.e., mangling a DLL name and change it on DLLs that depend on it
+
+_RG: Probably a gap indeed, not 100% sure. [Mach-O Mach-O mangler](https://github.com/njsmith/machomachomangler) supports Windows, but a Windows equivalent of [auditwheel](https://github.com/pypa/auditwheel) and [delocate](https://github.com/matthew-brett/delocate) seems to be missing._
+
+- This world does not know that the conda world exists.
+
+_RG: yeah, kinda. Devs do know conda and conda-forge exist, but as far as I can tell they do see the world as centered around PyPI, wheels and pip, and conda as one of many other package managers (next to Homebrew, Linux distros, Chocolatey, and so on). Which isn't quite right - PyData users now represent over half of all Python users, and the Conda ecosystem has specific capabilities related to that stack that are (and may remain) missing from PyPI/pip._
+
+- Shipping JupyterLab 3.0 extensions without Node.js is kinda cool.
+- Rust/Maturin to develop Python packages is nice.
+    - The relocation of depending binaries must be done manually in some cases.
+
+
+## topic 2: The conda world
+
+- Conda itself is almost unmaintained
+
+_RG: a quick peek at [the conda repo](https://github.com/conda/conda/pulse) indeed shows declining activity. Hopefully the new conda_
```

Comment:
I think this sentence needs to be completed.

Reply:
thanks, fixed